### ISSUE TITLE
Using the AVX-512 code path for Poly1305 clears non-volatile XMM registers and does not restore them.

### ISSUE THREAD

**Issue by zkWildfire** [openssl/openssl#21522]
When using the AVX-512 code path for poly1305, non-volatile XMM registers are cleared and not restored by the code executed by a call to the `poly1305_blocks` function pointer. This is similar to an issue previously reported (https://github.com/openssl/openssl/issues/12328), but that issue was for OpenSSL 1.1.1C and is marked as triaged. The version of OpenSSL I'm using which exhibits the buggy behavior is OpenSSL 3.1.1.

I've created a MCVE project here with further details: https://github.com/zkWildfire/OpenSslMcve  
Please see the MCVE project's README for how to set up the MCVE and for more information about the bug. Here's a short summary of the information covered in the README:

The MCVE project uses vcpkg to download and build OpenSSL. When building the project with no changes, the test app it builds will dump the XMM registers' state to the console and show how, when encrypting a short string, the bug does not get hit. When encrypting a longer string, registers XMM6-XMM15 get zeroed out by `poly1305_blocks()` and do not get restored before returning to the caller.

I've included a .patch file that can be applied via vcpkg that fixes the issue, though I'm not sure if the changes applied by the patch would be suitable for inclusion into OpenSSL directly. The project's readme has a section that covers what I did to fix the issue and a possible alternative to my approach. If the project's `vcpkg-configuration.json` is updated to point to `ports/fixed` instead of `ports/original`, clearing the build folder and rebuilding the project will demonstrate that the issue has been resolved by the patch.

This bug will only occur on processors that support the AVX-512F instruction set and are not based on Skylake. However, the MCVE can be used on any machine by using Intel's Software Development Emulator. I used both the emulator and a physical Intel Tiger Lake machine when investigating the bug and verifying my fix.

I've verified that my MCVE project can be set up from scratch on my local machine and on the Tiger Lake machine I used for testing. If issues are encountered when reproduce the poly1305 bug, please let me know and I'll see about fixing the issue.

**Comment by paulidale** [comment]
Any chance you could submit [the patch](https://github.com/zkWildfire/OpenSslMcve/blob/3d42bc1a133ad02764d1ebb7170af653080697d2/ports/fixed/openssl/windows/avx512.patch) as a pull request against OpenSSL?

**Comment by zkWildfire** [comment]
Yes, I can do that. However, since the patch was primarily meant as a reference and was focused on the specific code path tested in the MCVE, I'd want to do some more testing before I feel comfortable submitting it as something that can be included as-is. Would I be correct in assuming that running tests as described by [test/README.md](https://github.com/openssl/openssl/blob/master/test/README.md) is enough to verify that the fix doesn't break something else? I would be running the tests on both a processor that is unaffected by the bug as well as one that supports AVX-512 and is affected by the bug.

I can also include the code I wrote to check register state and write a unit test(s) to verify register state to prevent regressions if that would be helpful. However, since this bug only reproduces on specific processors, this might not be helpful if OpenSSL's CI/CD does not run tests on such a processor. Is this something that I should include in the pull request?

Lastly, are any concerns with going with the basic "push everything onto the stack and restore it later" approach I've taken? It may be possible to fix the issue by refactoring the code to avoid using the volatile XMM registers entirely. I'd have to get more familiar with the perl scripts used to generate the assembly code if I were to implement such a fix though.

**Comment by paulidale** [comment]
The instructions in test/README.md should be enough.  Note that you can disable individual processor features using the [OPENSSL_ia32cap](https://www.openssl.org/docs/manmaster/man3/OPENSSL_ia32cap.html) environment variable which would permit additional checking.  I'd suggest concentrating on using `evp_test`:
`./util/wrap evp_test test/recipes/30-test_evp_data/evpmac_poly1305.txt`
and `./util/wrap evp_test test/recipes/30-test_evp_data/evpciph_chacha.txt`
specifically.  This will let you test this algorithm in isolation.

The unit test to check the register state seems like an interesting addition.  It would require conditioning based on the CPU and it's capabilities (which might be painful).

I'm fine with pushing to the stack.  A lot of these assembly routines unroll loops and take full advantage of all the available registers to interleave iterations.  Still, if there is an unused register available, it wouldn't hurt to use it.

**Comment by zkWildfire** [comment]
Thanks! I'll poke around with validating my changes sometime in the next few days.

> The unit test to check the register state seems like an interesting addition. It would require conditioning based on the CPU and it's capabilities (which might be painful).

While the primary purpose of such a unit test would be to prevent regressions in the AVX-512 code path, the unit test(s) would be equally valid on non-AVX-512 processors (and could be valuable to validate/prevent regressions on the non-AVX-512 code paths too). Such functionality could be used to implement a register validation unit test for any OpenSSL method that drops into handwritten assembly code to avoid bugs like this from popping up again anywhere. For example, this functionality could be wrapped up in a method like this:

```c
bool ValidateRegisters(void(*testFunc)())
{
    // Set registers to some known state
    // Conditional compilation #ifdefs would also be found here as needed

    testFunc();

    // Validate register state and restore registers to their previous state, then
    //   notify the caller (presumably a test case) of the validation result so it can
    //   mark the test as failing if necessary
    return areRegistersValid;
}
```

For this specific bug, it could also be possible to update OpenSSL's CI/CD to run a very limited subset of the overall tests under Intel's Software Development Emulator to relax the requirement of needing an AVX-512 capable chip for validation. I wouldn't want to run many tests under the SDE since emulation is slow, but there would definitely be value in ensuring that the AVX-512 code path gets tested on CI/CD, especially if they aren't already.

As for the requirements, the unit test would be limited to checking XMM6-15 since those are the only non-volatile XMM registers. Pending verification, I believe that means the primary requirement would be that the processor supports SSE. However, the code would need to be conditioned on whether the current platform is Windows and whether the target architecture is x86 or x64. 

Does OpenSSL have a list of processor requirements listed anywhere? I glanced through a few .md files but I didn't see such a requirements list. If OpenSSL already has SSE as a minimum requirement, then the register verification unit tests would only need to account for the combination of Windows + x64. Or, if OpenSSL's build scripts already have some mechanism for conditionally compiling code based on specific targeted instruction sets, that could be leveraged.

> Still, if there is an unused register available, it wouldn't hurt to use it.

I'll take a look at how many instructions in the affected code path use a volatile XMM register and see how viable it would be to make that change instead. As it is, I only looked for where an affected register was used to determine where to place the save registers code block, so it's entirely possible that there are only a handful of instructions that would need to be changed.

**Comment by bernd-edlinger** [comment]
I think the 1.1.1 branch is affected as well.
May I suggest that you add similar labels like `.Ldo_avx512_body:` and `.Ldo_avx512_epilogue`,
And could you please try out if the exception handling is working, inside these functions?

**Comment by zkWildfire** [comment]
> I'll take a look at how many instructions in the affected code path use a volatile XMM register and see how viable it would be to make that change instead. As it is, I only looked for where an affected register was used to determine where to place the save registers code block, so it's entirely possible that there are only a handful of instructions that would need to be changed.

After taking a look at the perl script for generating the AVX-512 code path, I don't believe it's practical to revise the registers that are used to avoid using the non-volatile registers. The relevant lines of code are these:

```pl
my ($H0,$H1,$H2,$R0,$R1,$R2,$S1,$S2) = map("%ymm$_",(0..5,16,17));
my ($D0lo,$D0hi,$D1lo,$D1hi,$D2lo,$D2hi) = map("%ymm$_",(18..23));
my ($T0,$T1,$T2,$T3,$mask44,$mask42,$tmp,$PAD) = map("%ymm$_",(24..31));
my ($RR0,$RR1,$RR2,$SS1,$SS2) = map("%ymm$_",(6..10));
```
Source: [crypto/poly1305/asm/poly1305-x86_64.pl:3371](https://github.com/openssl/openssl/blob/master/crypto/poly1305/asm/poly1305-x86_64.pl#L3371)

All 32 YMM registers are used, so pushing XMM6-15 to the stack and restoring them later is the correct approach to take for fixing this bug.

> May I suggest that you add similar labels like .Ldo_avx512_body: and .Ldo_avx512_epilogue

I'm admittedly not quite sure what the best way to approach this would be. I know that the code already has similar labels for AVX and AVX2, but they're separate from the vpmadd52 labels. The vpmadd52 labels already come in different variants (e.g. `_4x` that requires AVX2, `_8x` that requires AVX-512). I don't think it would be a good idea to rename `vpmadd52_8x` to an AVX-512 label since it wouldn't be consistent with the other vpmadd52 labels then, and the other vpmadd52 labels (e.g. `vpmadd52_4x`) can't be renamed since the label names they'd be renamed to already exist. If you have a suggestion for how this can be done, I'm all ears.

> And could you please try out if the exception handling is working, inside these functions?

Could you elaborate on what you would like me to do? The changes I made to resolve this issue have been entirely in assembly. I admittedly don't have much experience with assembly, but as far as I know exceptions are something that are entirely handled by a compiler and are not something you can use from assembly code.

**Comment by bernd-edlinger** [comment]
> Could you elaborate on what you would like me to do? The changes I made to resolve this issue have been entirely in assembly. I admittedly don't have much experience with assembly, but as far as I know exceptions are something that are entirely handled by a compiler and are not something you can use from assembly code.

Certainly I did not mean to rename existing labels.
I would like you to try out if for instance an exception can be thrown from inside these vmpadd52_4x and/or vpmadd52_8x
and caught again outside.  That is what I think is the purpose of these code sections:

https://github.com/openssl/openssl/blob/bc5d9cc8711e86d5c25b81c58dfae531536e61fc/crypto/poly1305/asm/poly1305-x86_64.pl#L4111-L4115
https://github.com/openssl/openssl/blob/bc5d9cc8711e86d5c25b81c58dfae531536e61fc/crypto/poly1305/asm/poly1305-x86_64.pl#L4171-L4175

I think these infos are critical for exception handling inside the poly1305_blocks_avx512
function but I dont know if that is actually working or not (under linux it is actually verified to be working).
However if it is currently working it would be desireable to have similar SEH handlers for the vpmadd52_4x
and vpmadd52_8x sections if possible.

**Comment by zkWildfire** [comment]
I can definitely see how that would be worth adding. I won't be poking around with that until after the main bug is fixed though.

---

I've updated the MCVE code to check multiple other ciphers using the register checking code I wrote, then ran it under Intel's SDE and emulated a Tiger Lake CPU. I haven't tested all algorithms yet but so far the only affected cipher is chacha20_poly1305.

What I'm planning to do, pending approval from legal at work, is merge in my AVX-512 assembly patch and then add in code similar to the register code I've already written. The changes I made to my MCVE will be roughly mirrored in the added unit tests, which will allow register validation to be added to any OpenSSL method. I would also like to take a look at OpenSSL's CI/CD scripts and see how viable it would be to use Intel's SDE to selectively run only the register validation test cases so that issues like this bug would be detected by automated testing in the future.

**Comment by davidben** [comment]
> The unit test to check the register state seems like an interesting addition. It would require conditioning based on the CPU and it's capabilities (which might be painful).

We do this in BoringSSL and don't check in assembly functions without coverage with those testers. There have been a handful of ABI issues in OpenSSL assembly over the years, so I'd definitely recommend doing this. Paired with running under SDE in CI (which we also do), it's been immensely valuable in making sure all the subtle ABI details are correct. It's the same system I'd previously mentioned here:
https://github.com/openssl/openssl/issues/10470#issuecomment-557424564
https://github.com/openssl/openssl/pull/8109
https://github.com/openssl/openssl/pull/8108

We have testers for x86, x86_64, arm, and aarch64. At one point we also had [ppcle64](https://boringssl.googlesource.com/boringssl/+/31db68195bc1bb7542904b48ff8d770a8f920283%5E%21/), before we dropped that assembly. The surrounding machinery uses C++ ([header](https://boringssl.googlesource.com/boringssl/+/e5b6c141a/crypto/test/abi_test.h), [source](https://boringssl.googlesource.com/boringssl/+/e5b6c141a/crypto/test/abi_test.cc), [example use](https://boringssl.googlesource.com/boringssl/+/e5b6c141a/crypto/fipsmodule/sha/sha_test.cc)) and also includes support for unwind testing, which is a bit involved. You all probably just want the register-checking one, which is much smaller and should be easy to adapt to your C-based tests. Most of `abi_test.cc` is the unwind tester for checking CFI and SEH bits, and is somewhat in progress. (Most of OpenSSL's Windows assembly is incorrect w.r.t. SEH unwind.) The [assembly trampolines](https://boringssl.googlesource.com/boringssl/+/e5b6c141a/crypto/test/asm/) can probably be imported with minimal changes.

(You're welcome to treat these as a contribution under our CLA.)

**Comment by t8m** [comment]
@zkWildfire @bernd-edlinger

This issue was published as CVE-2023-4807 Low severity security issue on Friday 8th September 2023.

https://www.openssl.org/news/secadv/20230908.txt

**Comment by zkWildfire** [comment]
Hi @bernd-edlinger @t8m, apologies for dropping off the radar for a bit. It took my work around a month to get everything sorted on the legal side of things and get the corporate CLA taken care of. I emailed both the corporate CLA and an individual CLA for myself to [legal@opensslfoundation.org](mailto:legal@opensslfoundation.org), which I found on OpenSSL's site here:
https://www.openssl.org/policies/cla.html

I asked for a reply confirming that I can work on OpenSSL now but I haven't heard back. Is sending the CLAs off enough or do I need confirmation from someone on OpenSSL's legal team before I can begin work? I emailed the CLAs to OpenSSL around two weeks ago now.
